plotter/contour.go

```diff
@@ -0,0 +1,659 @@
+				min = math.Min(min, v)
+				max = math.Max(max, v)
+			}
+		}
+	}
+
+	if len(levels) == 0 {
+		levels = quantilesR7(g, defaultQuantiles)
+	}
+
+	return &Contour{
+		GridXYZ:    g,
+		Levels:     levels,
+		LineStyles: []draw.LineStyle{DefaultLineStyle},
+		Palette:    p,
+		Min:        min,
+		Max:        max,
+	}
+}
+
+// Default quantiles for case where levels is not explicitly set.
+var defaultQuantiles = []float64{0.01, 0.05, 0.25, 0.5, 0.75, 0.95, 0.99}
+
+// quantilesR7 returns the pth quantiles of the data in g according the the R-7 method.
+// http://en.wikipedia.org/wiki/Quantile#Estimating_the_quantiles_of_a_population
+func quantilesR7(g GridXYZ, p []float64) []float64 {
+	c, r := g.Dims()
+	data := make([]float64, 0, c*r)
+	for i := 0; i < c; i++ {
+		for j := 0; j < r; j++ {
+			if v := g.Z(i, j); !math.IsNaN(v) {
+				data = append(data, v)
+			}
+		}
+	}
+	sort.Float64s(data)
+	v := make([]float64, len(p))
+	for j, q := range p {
+		if q == 1 {
+			v[j] = data[len(data)-1]
+		}
+		h := float64(len(data)-1) * q
+		i := int(h)
+		v[j] = data[i] + (h-math.Floor(h))*(data[i+1]-data[i])
+	}
+	return v
+}
+
+// naive is a debugging constant. If true, Plot performs no contour path
+// reconstruction, instead rendering each path segment individually.
+const naive = false
+
+// Plot implements the Plot method of the plot.Plotter interface.
+func (h *Contour) Plot(c draw.Canvas, plt *plot.Plot) {
+	if naive {
+		h.naivePlot(c, plt)
+		return
+	}
+
+	var pal []color.Color
+	if h.Palette != nil {
+		pal = h.Palette.Colors()
+	}
+
+	trX, trY := plt.Transforms(&c)
+
+	// Collate contour paths and draw them.
+	//
+	// The alternative naive approach is to draw each line segment as
+	// conrec returns it. The integrated path approach allows graphical
+	// optimisations and is necessary for contour fill shading.
+	cp := contourPaths(h.GridXYZ, h.Levels, trX, trY)
+
+	// ps is a palette scaling factor to scale the palette uniformly
+	// across the given levels. This enables a discordance between the
+	// number of colours and the number of levels. Sorting is not
+	// necessary since contourPaths sorts the levels as a side effect.
+	ps := float64(len(pal)-1) / (h.Levels[len(h.Levels)-1] - h.Levels[0])
+	if len(h.Levels) == 1 {
+		ps = 0
+	}
+
+	for i, z := range h.Levels {
+		if math.IsNaN(z) {
+			continue
+		}
+		for _, pa := range cp[z] {
+			if isLoop(pa) {
+				pa.Close()
+			}
+
+			style := h.LineStyles[i%len(h.LineStyles)]
+			var col color.Color
+			switch {
+			case z < h.Min:
+				col = h.Underflow
+			case z > h.Max:
+				col = h.Overflow
+			case len(pal) == 0:
+				col = style.Color
+			default:
+				col = pal[int((z-h.Levels[0])*ps+0.5)] // Apply palette scaling.
+			}
+			if col != nil && style.Width != 0 {
+				c.SetLineStyle(style)
+				c.SetColor(col)
+				c.Stroke(pa)
+			}
+		}
+	}
+}
+
+// naivePlot implements the a naive rendering approach for contours.
+// It is here as a debugging mode since it sinply draw line segments
```

Comment:
s/sinply/simply

Something is wrong with "draw line segments returned generated"